Production/scripts/validation_tool.py

```diff
@@ -0,0 +1,159 @@
+  'tau_eta'   : (5, -2.5, 2.5),
+  'lepton_gen_match' : (20, -1, 19),
+  'sampleType': (20, -1, 19),      
+  'dataset_id': (20, -1, 19),
+  'dataset_group_id': (20, -1, 19),
+}
+
+def groupby(dataframe, by):
+  _ = dataframe.Histo1D(by)
+  hist = _.GetValue()
+  hist.ClearUnderflowAndOverflow()
+  types = list(set([round(hist.GetBinCenter(jj)) for jj in range(hist.GetNbinsX()) if hist.GetBinContent(jj)]))
+  types = [int(tt) for tt in types]
+
+  return {tt: dataframe.Filter('{} == {}'.format(by, tt)) for tt in types}
+
+def get_histos(dataframe, branch, norm = False):
+  size = int(dataframe.Count())
+  sub_size = 1 + size / N_SPLITS
+  subframes = [dataframe.Range(ii*sub_size, (ii+1)*sub_size) for ii in range(N_SPLITS)]
+  
+  model = (branch, "") + BINS[branch]
+  hptrs = [sf.Histo1D(model, branch) for sf in subframes]
+  histos = [hh.GetValue().Clone() for hh in hptrs]
+  
+  for ii, hh in enumerate(histos):
+    hh.SetTitle(branch)
+    hh.SetName('_'.join([hh.GetName(), ii]))
+    hh.Sumw2()
+    hh.ClearUnderflowAndOverflow()
+    if norm and hh.Integral():
+      hh.Scale(1. / hh.Integral())
+
+  return histos
+
+def save_histos(histos, fdir, pvalues):
+  OUTPUT_ROOT.cd()
+  OUTPUT_ROOT.cd(fdir)
+  
+  can = ROOT.TCanvas()
+  leg = ROOT.TLegend(0.9, 0.1, 1., 0.9, "p-values (KS with the first chunk)")
+
+  histos[0].GetYaxis().SetRangeUser(0, 1.1*max(hh.GetMaximum() for hh in histos))
+  histos[0].SetMarkerStyle(20)
+
+  for ii, hh in enumerate(histos): 
+    hh.SetLineColor(ii+1)
+    hh.SetMarkerColor(ii+1)
+    leg.AddEntry(hh, 'chunk %d - pval = %.3f' %(ii, pvalues[ii]), 'lep')
+    hh.Draw("PE" + " SAME" * (ii != 0))
+    hh.Write()
+  if args.legend:
+    leg.Draw("SAME")
+  if args.pdf is not None:
+    can.SaveAs('{}/{}.pdf'.format(args.pdf, fdir.replace('/', '_')), 'pdf')
+  can.Write()
+  OUTPUT_ROOT.cd()
+
+def run_validation(dataframe, branches, pwd):
+  OUTPUT_ROOT.cd()
+  OUTPUT_ROOT.mkdir(pwd)
+  
+  if not pwd in JSON_DICT.keys():
+    JSON_DICT[pwd] = OrderedDict()
+
+  for branch in branches:
+    OUTPUT_ROOT.cd() 
+    OUTPUT_ROOT.mkdir('/'.join([pwd, branch]))
+
+    histos = get_histos(dataframe, branch = branch, norm = True)
+
+    pvalues = [histos[0].KolmogorovTest(hh) if histos[0].Integral()*hh.Integral() else 99 for hh in histos]
+    if not histos[0].Integral():
+      print '[WARNING] control histogram is empty for step {} inside {}'.format(branch, pwd)
+    
+    if not all([pv >= PVAL_THRESHOLD for pv in pvalues]):
+      print '[WARNING] KS test failed for step {} inside {}. p-values are:'.format(branch, pwd)
+      print '\t', pvalues
+    
+    JSON_DICT[pwd][branch] = pvalues
+    
+    save_histos(histos, fdir = '/'.join([pwd, branch]), pvalues = pvalues)
+
+if __name__ == '__main__':
+  print '[INFO] reading files', args.input
+  input_files = ROOT.std.vector('std::string')()
+  for file in glob.glob(args.input):
+    input_files.push_back(str(file))
+
+  main_dir = 'KS_test'
```

Comment:
I suggest to use `args.output` as the main dir and store output root file inside it.

Reply:
hm.. actually I misread the code, `main_dir` is used for the path inside the root file. Ignore the previous comment.